### ISSUE TITLE
sttp2: allow relative URIs

### ISSUE THREAD

**Issue by kamilkloch** [softwaremill/sttp#285]
Right now the host is always required. In the scala.js context it is desireable to operate with relative uris, just like a standard Javascript `Ajax.get("/srv")` call on the client side in the browser.

Host would need to become optional.

**Comment by adamw** [comment]
To do this properly, we would need to introduce a concept of a URI reference: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#URI_references - probably a new type.

That would also mean a new parser (interpolator).

Then the request would contain an `Either[UriRef, Uri]`, instead of the `Uri` field now.

However, I'm wondering if this would make sense for non-js backends: they all require an `Uri`, not a reference, as they have no idea as to how handle references. So I'm not sure if the above solution would in fact do more good than harm.

An alternative - however quite hacky - would be to modify the `FetchBackend` to remove the scheme + authority if the host is equal to e.g. `"local"`. I know - that's a magic string and doesn't look good. Also, it wouldn't allow differentiating between path-relative (e.g. `"a/b/c"`) and path-absolute (e.g. `"/a/b/c"`) requests.

Any ideas? :)

**Comment by aappddeevv** [comment]
Can you do this by just creating a baseline request object and adding a "path" method that adds to the path, returning a new request, vs replacing it. Then passing that around as a copyable object?

**Comment by kamilkloch** [comment]
@aappddeevv I like the idea. Simple to implement. We could even leave `path` unchanged and add `/` (already existing in http4s or tapir). So it would look like this:
```scala
val root = uri"http://localhost/x/y"
val relative = root / ".." / "z" // http://localhost/x/z
```
This would still not solve all issues (relative uris automatically follow http/https protocol) but it's a nice start :)

**Comment by aappddeevv** [comment]
You could just have a method that adds the path segment and changes the protocol. The only downside of the approach is you have to pass around a request object as the "root context." But it does make the syntax more path'ish. I think there was a scala URI-only library and a "trail" lib that had something like for constructing paths. scalajs-react also does something like this. So these could all be inspirations.

**Comment by aappddeevv** [comment]
You could also allow your backend to take a "baseline" request object that is "combined" in an applicative sense with the request being sent so you don't need to pass around a root. It's not a commutative monoid though.

**Comment by adamw** [comment]
@kamilkloch not sure I understand - would that be different from `uri"$root/../z` (assuming the `..` work in uris, which I'm not sure, but if not, we should fix that ;) )? Or just different syntax?

**Comment by adamw** [comment]
@aappddeevv not sure I understand - how would that solve the original problem? Maybe you could add some code examples of how this could look like?

**Comment by aappddeevv** [comment]
Using the root idea and a "/" combinator is as described above. For the baseline request, you would just do (totally schematic code):
```scala
class SttpBackend(blahParameter: String, baseRequest: RequestT[_]) { 
  def prep(r: RequestT[T]) = baseRequest combine r
}
```
where you have defined your combine method somewhere or its a typeclass, etc. Essentially, it gets rid of the need to pass around a "root" request object. A bunch of dedicated js clients have this so that they can discover the proper API URL and users just have to worry about the relative paths. Or they pass in a root path instead of a base request.

**Comment by kamilkloch** [comment]
> @kamilkloch not sure I understand - would that be different from `uri"$root/../z"` (assuming the `..` work in uris, which I'm not sure, but if not, we should fix that ;) )? Or just different syntax?

@adamw `uri"$root/../z"` is working fine :) You are right, the functionality is here, personally I lean towards the slash syntax, it just reads nicer (e.g. in tapir: `.in("srv" / "worker" / "status")`), but this is cosmetics. If you feel the slashes would (additionally) fit, I am happy to make a PR. 

Coming back to the main topic: introducing separate `UriRef` complicates the architecture indeed. As for the magic `local` string, this gets even more confusing with the scheme (in the browser `x/y` will automatically pick the scheme from the web page, so writing `http://local/x/y` is a lie - in practice, it can be https). From those two, I lean towards `UriRef`. I do not have a better proposition :/

**Comment by adamw** [comment]
@kamilkloch yeah, I'm going back and forth on the `/` syntax. On one hand, it is nice and natural, on the other, it opens up the door to `?` syntax (or similar) for specifying query parameters. And it would mean there are two different ways of specifying the path (through interpolation and this new syntax). So ... I don't know :)

Assuming we would introduce a `UriRef`, what would be the behavior of non-js backends if you'd want to send a request with such an uri?

**Comment by aappddeevv** [comment]
Why do you need a UriRef in the uri"$root/..." scheme? Can you just add a combinator that combines two Uris?

**Comment by adamw** [comment]
@aappddeevv two styles? there are two things here: one is a way of making a uri more specific - adding path components, or query parameters to an existing uri. Currently you can do it like this: `uri"$existingUri/a/b/c?x=y"`, Kamil was proposing adding syntax like this: `existingUri / "a" / "b" / "c"`.

The second problem are relative uris. In JS the application often doesn't know the base uri at all, it only does requests relative to the root path. The browser adds the schema/domain automatically

**Comment by adamw** [comment]
@kamilkloch yet another solution would be to relax the requirements of `Uri` and simply make it accept uri-references as well, that is allowing optional/empty schemes/domains

**Comment by kamilkloch** [comment]
@adamw I guess sensible defaults for scheme & host would be `http://localhost`. For what it's worth, `curl` does default to `http` but does not default to `localhost`. Perhaps such defaults would not be necessary, at all though. Rough idea: what do you think about adding `UriRef` contructors __only__  for JS? (`def get/post/put/delete(uri: UriRef): Request[T, S] = ...`). Ideally, `RequestT` would require `Uri` on JVM and allow for `Uri` or `UriRef` in JS :)

**Comment by adamw** [comment]
This would cause code to compile either for JS or JVM, and could be quite surprising to users :)

I'm leaning towards simply making the scheme, host and ports optional in the `Uri` class

**Comment by kamilkloch** [comment]
Any plans of baking it in to v2? Should I try to chip in?

**Comment by adamw** [comment]
@kamilkloch I still don't have a good idea on how to implement this so that the JVM version works well, but the original JS problem is also fixed.

I think I'd like to keep `Uri` to represent a valid URI - not a relative one. Ideas/PRs always welcome :)

**Comment by francisdb** [comment]
Something like this would be useful

https://github.com/http4s/http4s/blob/037ea813dd5efeae0995492b573d67e22a8d8fb8/core/src/main/scala/org/http4s/Uri.scala#L816

**Comment by akkie** [comment]
I also have the need for relative URIs on the JVM. I've created https://github.com/softwaremill/sttp-model/issues/12 a while ago but got no response. For [Silhouette](https://github.com/minutemen/silhouette/blob/master/modules/provider-oauth2/src/main/scala/silhouette/provider/oauth2/OAuth2Provider.scala#L384) I allow to define callback URIs. These can be absolute or relative. Relative URIs will then be resolved against the current host.

In my opinion the STTP model implementation should provide this functionality because relative URIs are defined in RFC 3986.

I understand that on the JVM it makes no sense to send a request to a relative URI. So maybe it makes sense to implement methods like ` Uri.isRelative` or `Uri.isAbsolute`. Before a request is send to an URI, STTP could check if the URI is absolute and if not it throws an exception.

**Comment by adamw** [comment]
@akkie I understand the problem and some of the use-cases, but I still don't know how to best implement this :)

First we have the question - what should the `Uri` class represent? If we are conservative and stick to RFC3986 4.1:

```
   URI-reference is used to denote the most common usage of a resource
   identifier.

      URI-reference = URI / relative-ref

   A URI-reference is either a URI or a relative reference.  If the
   URI-reference's prefix does not match the syntax of a scheme followed
   by its colon separator, then the URI-reference is a relative
   reference.
```

then we would need two types: `Uri` and `UriRelative`/`RelativeUri`. As I understand, a `URI` as defined there is the full form (with the scheme), while a relative one allows this part to be skipped.

But maybe it would be pragmatic to say that the `Uri` class in sttp-model corresponds to what is defined as `URI-reference` in RFC? That would amount to making the scheme and host optional + adding a resolve method (probably not that trivial to implement, but can be a nice task :) ). 

That would also mean that trying to send a request with a relative uri using a normal backend would end with an exception - exceptions are generally not that great to encounter during normal use, but again maybe that's the pragmatic choice.

Any opinions on which route to take? :)

**Comment by akkie** [comment]
@adamw Good question. The Java URI implementation and also the linked http4s `Uri` implementation supports the functionality in one class. But providing `UriRelative`/`RelativeUri` would fix the exception issue.

According to my gut feeling, I would go the pragmatic way.

**Comment by adamw** [comment]
@francisdb @kamilkloch @aappddeevv @ghostbuster91 @LGLO opinions? :)

**Comment by francisdb** [comment]
+1 for the pragmatic solution of not having 2 Uri instances

Also looking for something like this:
https://w3lib.readthedocs.io/en/latest/w3lib.html#w3lib.url.canonicalize_url
but that's probably for a different ticket

Maybe everything url related could be a separate lib? Looks like all http libs in the scala space have their own implementation.
https://doc.akka.io/api/akka-http/current/akka/http/scaladsl/model/Uri.html
And some stick to `java.net.URI` with utilities

**Comment by kamilkloch** [comment]
Nice to see the topic being revived :) Also +1 for the optional scheme and host (if this is kind of change is now really on the table). Perhaps this would even allow for interop with `Uri` from http4s :)

**Comment by ghostbuster91** [comment]
In my opinion this should be a separate type like path class in os-lib by lihaoyi https://github.com/lihaoyi/os-lib/blob/master/os/src/Path.scala#L240 

I am aware that it will require more work to add support for relative URIs using that approach, but in the long run it is a small price to pay for having consistent, compile-time verifiable code. 

Don't fall for the easy way path just because it is easy.

**Comment by adamw** [comment]
@ghostbuster91 one of the original problems remain, though: we want to create a `Request` with both relative & absolute URIs. Though only JS backends can send requests using relative uris, while jvm backends can only send request using absolute ones.

But maybe ... we could try and leverage capabilities here? If we had a capability `Relative`, that would allow sending requests with relative uris. We could then create a wrapper backend which would lift any other backend to a relative-capable one, by providing a base address against which relative uris would be resolved.

**Comment by adamw** [comment]
After some research I've taken the route where `Uri` can be both relative & absolute. No capabilities or subtypes.

First of all, if we really want to model a `Uri`, everything except for the schema needs to be optional (see https://tools.ietf.org/html/rfc3986). That's the only required component which makes a URI absolute. Of course, a URI without an authority doesn't make much sense in the http world - so if we wanted to model the hierarchy exactly, we would need not only an `Uri`, `RelativeUri` but also a `HttpUri` (or simply `Url`) classes. This would complicate things quite a lot without much obvious benefit.

Secondly, differentiating between relative & absolute requests isn't that easy; capabilities wouldn't really work here due to variance & subtyping (the ability to case an absolute request to a relative one). Adding a second method to `SttpBackend` isn't possible as most backends don't support relative URIs anyway.

In the end, the `Uri` class represents what the RFC calls a URI reference - either an absolute or relative uri. If a wrong one is passed to a backend, an exception will be thrown - which is a sad fact, but one with which we have to live. But we can't eliminate exceptions as uris can have incorrect schemes or hosts, even if all required components are there.

See the PR: https://github.com/softwaremill/sttp-model/pull/51, which changes the `Uri` class so that authority & scheme are optional, and adjusts the uri interpolator accordingly.

**Comment by adamw** [comment]
Should be fixed in 3.0.0-RC14. There's a `ResolveRelativeUrisBackend` backend wrapper, which allows to specify the base uri.

Otherwise backends don't work with relative URIs - they will probably throw exceptions when given a relative one. Although the fetch backend in a browser env should work as far as I understand.

**Comment by kamilkloch** [comment]
It works indeed, thank you @adamw!
